common/io/Serial.cpp

```diff
@@ -55,17 +123,81 @@ bool UIntToSpeedT(uint32_t value, speed_t *output) {
   return false;
 }
 
-bool CheckForUUCPLockFile(const std::vector<std::string> &directories,
-                          const std::string &serial_device) {
-  vector<string>::const_iterator iter = directories.begin();
-  for (; iter != directories.end(); ++iter) {
-    const string lock_file = (
-        *iter + ola::file::PATH_SEPARATOR + "LCK.." + serial_device);
-    if (FileExists(lock_file)) {
-      return true;
+
+bool AcquireUUCPLockAndOpen(const std::string &path, int oflag, int *fd) {
+  // This is rather tricky since there is no real convention for LCK files.
+  // If it was only a single process doing the locking we could use fnctl as
+  // described in 55.6 of the Linux Programing Interface book.
+
+  // First, clean up a stale lockfile.
+  const string lock_file = GetLockFile(path);
+  OLA_DEBUG << "Checking for " << lock_file;
+  pid_t locked_pid;
+  if (!GetPidFromFile(lock_file, &locked_pid)) {
+    return false;
+  }
+
+  if (locked_pid) {
+    // This will return false even if we have the lock, this is what we want
+    // since different plugins may try to open the same serial port - see issue
+    // #888.
+    if (ProcessExists(locked_pid)) {
+      return false;
+    }
+    // There is a race between the read & the unlink here. I'm not convinced it
+    // can be solved.
+    if (!RemoveLockFile(lock_file)) {
+      return false;
+    }
+  }
+
+  pid_t our_pid = getpid();
+  // Now try to open the file in O_EXCL mode. If we fail, someone else already
+  // took the lock.
+  int lock_fd = open(lock_file.c_str(), O_RDWR | O_CREAT | O_EXCL,
+                     S_IRUSR | S_IWUSR | S_IRGRP | S_IWGRP | S_IROTH | S_IWOTH);
+  if (lock_fd < 0) {
+    return false;
+  }
+
+  OLA_INFO << "Acquired " << lock_file;
```

Comment:
the name contains LCK, so it's clear it's a lock file. I'm trying to reduce logging output